app/controllers/contests_controller.rb

```diff
@@ -0,0 +1,241 @@
+                  notice: "Contest was successfully ended."
+    else
+      render :admin, status: :unprocessable_entity
+    end
+  end
+
+  # POST /contest/create
+  # rubocop:disable Metrics/MethodLength
+  def create
+    authorize Contest, :admin?
+
+    if concurrent_contest_exists?
+      redirect_to contests_admin_path,
+                  notice: "Concurrent contests are not allowed. Close other contests before creating a new one."
+      return
+    end
+
+    @contest = Contest.new(
+      contest_params.reverse_merge(
+        deadline: 1.month.from_now,
+        status: :live
+      )
+    )
+
+    if @contest.save
+      User.find_each(batch_size: 1_000) do |user|
+        ContestNotification.with(contest: @contest).deliver_later(user)
+      end
+      redirect_to contest_page_path(@contest),
+                  notice: "Contest was successfully started."
+    else
+      render :admin, status: :unprocessable_entity
+    end
+  end
+  # rubocop:enable Metrics/MethodLength
+
+  # GET /contests/:id/new_submission
+  def new_submission
+    @projects   = current_user.projects
+    @contest    = Contest.find(params[:id])
+    @submission = Submission.new
+  end
+
+  # POST /contests/:id/create_submission
+  def create_submission
+    project_id = params[:submission][:project_id]
+
+    # 1️⃣ Ensure the project belongs to the current user
+    return redirect_unauthorized_project unless project_owner?(project_id)
+
+    # 2️⃣ Prevent duplicate submissions of the same project to the contest
+    return redirect_duplicate_submission if duplicate_submission?(project_id)
+
+    # 3️⃣ Create the submission
+    @submission = Submission.new(
+      project_id: project_id,
+      contest_id: params[:contest_id],
+      user_id: current_user.id
+    )
+
+    if @submission.save
+      redirect_to contest_page_path(params[:contest_id]),
+                  notice: "Submission was successfully added."
+    else
+      render :new_submission, status: :unprocessable_entity
+    end
+  end
+
+  # PUT /contests/:contest_id/withdraw/:submission_id
+  def withdraw
+    submission = Submission.find(params[:submission_id])
+
+    # Only the owner or an admin can withdraw a submission
+    unless submission.user_id == current_user.id || current_user&.admin?
+      redirect_to contest_page_path(params[:contest_id]),
+                  alert: "You are not allowed to withdraw this submission." and return
+    end
+
+    submission.destroy!
+    redirect_to contest_page_path(params[:contest_id]),
+                notice: "Submission was successfully removed."
+  end
+
+  # POST /contests/:contest_id/submission/:submission_id/upvote
+  # rubocop:disable Metrics/MethodLength
+  def upvote
+    contest = Contest.find(params[:contest_id])
+    redirect_to contest_page_path(contest), alert: "Voting is closed." and return if contest.completed?
+
+    user_votes = current_user.user_contest_votes(contest.id)
+
+    notice = if user_votes >= 3
```

Comment:
Move this to constant in SubmissionVote class
```suggestion
    notice = if user_votes >= SubmissionVote::USER_VOTES_PER_CONTEST
```